### PR TITLE
Porting Add Missing References Tests

### PR DESCRIPTION
**Pull request by jmarolf**
depends on https://github.com/dotnet/roslyn/pull/17431 being merged

### CODE REVIEW

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs [hunk 1]
```diff
@@ -161,7 +235,7 @@ public void CleanUpOpenSolution()
 
                 // Save the full path to each project in the solution. This is so we can
                 // cleanup any folders after the solution is closed.
-                foreach (EnvDTE.Project project in dte.Solution.Projects)
+                foreach (var project in dte.Solution.Projects.Cast<EnvDTE.Project>())
```
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Why?

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs [hunk 2]
```diff
@@ -189,10 +263,57 @@ private EnvDTE.Project GetProject(string nameOrFileName)
                 => string.Compare(p.FileName, nameOrFileName, StringComparison.OrdinalIgnoreCase) == 0
                 || string.Compare(p.Name, nameOrFileName, StringComparison.OrdinalIgnoreCase) == 0);
 
-        public void AddFile(string projectName, string fileName, string contents = null, bool open = false)
+        public void UpdateOrAddFile(string projectName, string fileName, string contents = null, bool open = false)
```
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: The `UpdateOrAddFile` / `UpdateFile` / `AddFile` methods are all screaming for some doc comments.

File: src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs [hunk 3]
```diff
@@ -189,10 +263,57 @@ private EnvDTE.Project GetProject(string nameOrFileName)
...
+
+        public void UpdateFile(string projectName, string fileName, string contents = null, bool open = false)
+        {
+            void SetText(string text)
+            {
+                InvokeOnUIThread(() =>
+                {
+                    // The active text view might not have finished composing yet, waiting for the application to 'idle'
+                    // means that it is done pumping messages (including WM_PAINT) and the window should return the correct text view
+                    WaitForApplicationIdle();
```
Comment (tmeschter) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Is this something we should be concerned about in other places?

Comment (jmarolf) [inline src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Nope this is handled in the InProc Editor APIs already

**Pull request merged**